Add support for animation via ngAnimate

lexi-lambda opened the pull request:
Fairly self-explanatory: this just uses `$animate` for DOM manipulation to make use of Angular's animation hooks. It also includes a use of the animation in the "custom" sample.

I also replaced the direct usage of `$templateCache` with `$templateRequest`, which does exactly what was already being done, anyway.

dwmkerr commented:
Hi @lexi-lambda this looks great. Two questions:

1. What happens if someone uses the modal service but *doesn't* have ngAnimate installed?
2. There's a new `closeComplete` part of the API. Should this in fact *replace* close? If there are valid use cases when clients could use either, maybe we could provide a `closing` and `closed` promise instead.

WDYT? If we have to include `ngAnimate` when using the modal service this could be problematic. Otherwise, it's a great addition to the library and I'd love to get it in.

lexi-lambda commented:
Sorry for the late response. Let me address your questions:

1. This is the neat thing about ngAnimate: Angular has an `$animate` service built-in, even without ngAnimate installed. The built-in service doesn't actually do any animation, it just does DOM manipulation, so it will degrade gracefully. If ngAnimate *is* installed, then `$animate` is decorated with animation functionality.
2. I think there *may* potentially be valid use-cases for either one, but I'm not actually sure about that. This is a valid point to discuss. I don't think it would be unreasonable for `closeComplete` to replace `close`, but that's also making some assumptions about users' use-cases that might not be valid.

dwmkerr commented:
Hi @lexi-lambda,

1. Brilliant - let's get this in ASAP!
2. You are right. What I'm thinking is that maybe what we should do is leave `closeComplete` as an **additional** part of the API (as it's non contract breaking) and document it. However, perhaps as a major release we should consider that we move to:
- `closing`: The user is closing the modal. Called immediately when the user confirms closure, runs before animations. Useful for business logic.
- `closed`: All animations etc are completed and the modal is completely gone. Useful for UI logic.

WDYT?

If you think this is a good way to go can you update your PR from master so I can merge directly in?

lexi-lambda commented:
Okay, I changed `close` → `closing` and `closeComplete` → `closed`, and I've rebased off of master. Look good to you? I'll leave it up to you to bump the version.

dwmkerr commented:
Hi @lexi-lambda can you leave `closing` as `close` now so as not to break the API, then in a separate branch I'll start on the next major release and rename the API there. That way this PR is only two features:

1. ngAnimate support
2. new promise `closed` (but existing APIs remain exactly the same)

Thanks!

lexi-lambda commented:
Done: `closing` is once again just called `close`.

dwmkerr commented on dst/angular-modal-service.js:
```diff
@@ -28,11 +28,10 @@
-            .then(function(result) {
-              deferred.resolve(result.data);
-            }, function(error) {
+          $templateRequest(templateUrl, true)
```
nice :smile:

dwmkerr commented:
This is a fantastic contribution @lexi-lambda I'm merging now and will release later today, I'll also add you to the contributor list, thanks so much!

StormPooper commented:
Any word on when this will be released? I've had to point to the latest commit to get this working in our project in the meantime, but I'd like to be able to use a stable release version.

bsitruk commented:
Hello,
Same question here. When will this be released ?
